02-oop-inheritance/keypoints/2_7-12/script.js

```diff
@@ -0,0 +1,150 @@
+    childClass.prototype = copyOfParent;
+};
+
+//MovieObserver can listen one or more actions like play or stop
+var MovieObserver = function (listenTo) {
+    this.listenTo = listenTo || [] ;
+};
+
+MovieObserver.prototype.play = function (movie) {
+    console.log('Playing: ' + movie.title);
+};
+
+MovieObserver.prototype.stop = function (movie) {
+    console.log('Stopped: ' + movie.title);
+};
+
+MovieObserver.prototype.download = function (movie) {
+    console.log('Downloading: ' + movie.title);
+};
+
+//The observerList can publish the actions in your own list of observer
+var ObserverList = function () {
+    this.observerList = [];
+};
+
+ObserverList.prototype.publish = function (action) {
+    var observerIndex = 0;
+
+    for (observerIndex ; observerIndex < this.observerList.length ; observerIndex++) {
+        for (var listenToIndex = 0 ; listenToIndex < this.observerList[observerIndex].listenTo.length ; listenToIndex++) {
+            if (this.observerList[observerIndex].listenTo[listenToIndex] === action) {
+                if (action === 'play') {
+                    this.observerList[observerIndex].play(this);
+                } else if (action === 'stop') {
+                    this.observerList[observerIndex].stop(this);
+                } else if (action === 'download') {
+                    this.observerList[observerIndex].download(this);
+                }
+            }
+        }
+    }
+};
+
+ObserverList.prototype.subscribe = function (observer) {
+    this.observerList.push(observer);
+};
+
+ObserverList.prototype.unsubscribe = function (observer) {
+    var observerIndex = 0;
+
+    for (observerIndex ; observerIndex < this.observerList.length ; observerIndex++) {
+        if (this.observerList[observerIndex] === observer) {
+            this.observerList.slice(observerIndex, 1);
+            break;
+        }
+    }
+};
+
+//Exercise 7
+var Movie = ( function () {
+
+    function Movie(title, genre, rating) {
+
+        ObserverList.call(this);
+
+        this.title = title || '';
+        this.rating = rating || 0;
+        this.genre = genre || '';
+
+        console.log('A movie called ' + title + ' was created');
+    };
+
+    inheritPrototype(Movie, ObserverList);
+
+    Movie.prototype.get = function (key) {
+        return this[key];
+    };
+
+    Movie.prototype.play = function () {
+        this.publish('play');
+    };
+
+    Movie.prototype.set = function (key, value) {
+        this[key] = value;
+    };
+
+    Movie.prototype.stop = function () {
+        this.publish('stop');
+    };
+
+    return Movie;
+
+})();
+
+//Exercise 8
+var DownloadableMovie = function (title, genre, rating) {
+    Movie.call(this, title, genre, rating);
+};
+
+inheritPrototype(DownloadableMovie, Movie);
+
+DownloadableMovie.prototype.download = function () {
+    this.publish('download');
+};
+
+//Exercise 9
+var Social = function () {
+
+};
+
+
+Social.prototype.share = function (friendName) {
+
```

Comment:
I guess I should imagine what this function does :P hahaha. No need of change for this exercise, try to be aware of this for next exercise.